### ISSUE TITLE
Tweak Map

### ISSUE THREAD

**Issue by LeonYang5114** [imagej/imagej-ops#313]
The newly merged branch #311 has some issues as indicated by @dietzc and @ctrueden, and I summarize them here so that we could work on fixing it in the hackathon or later.

- [x] Naming. We agreed on calling `RAI` for `RandomAccessibleInterval` but not `II` for `IterableInterval`, in naming ops. However, using `IterableInterval` when naming `Map` ops will generate names such as `MapIterableIntervalAndIterableIntervalOutputIterableInterval`. (This issue also exists in the math template ops [IIToIIOutputII](https://github.com/imagej/imagej-ops/blob/master/src/main/templates/net/imagej/ops/math/IIToIIOutputII.vm). Names such as `MapTwoIterableIntervalToIterableInterval` had been suggested.

- [ ] Avoid method calls within loops. For example [here](https://github.com/imagej/imagej-ops/blob/master/src/main/java/net/imagej/ops/map/MapIIAndIIToRAI.java#L81). The `op` can be retrieved outside the loop.

- [x] MapBinaryInplace. Do we actually need this interface and its implementations? `Map`s extending this interface are supposed to support mutation of either `in1()` or `in2()`, but this introduce some type unsafety, if the current implementation is not changed. We could probably remove this interface, and any binary `Map` extending `MapInplace` mutates the first input by default.

- [x] MapBinaryComputer implementations not complete. Right now we have 3 binary `Map` ops:
II+II=>II, II+II=>RAI, II+RAI=>RAI. To make the collection complete, we still need
II+RAI=>II, RAI+II=>II, 
RAI+II=>RAI, RAI+RAI=>II
(RAI+RAI=>RAI is not necessary since one of the RAIs could be converted into II)
The problem is, this proliferates the namespace, and introduces ambiguity. Do we want more of these binary `Map` ops? or do we want to remove some we currently have? or right now is fine?

More issue will be added later.

**Comment by ctrueden** [comment]
> MapBinaryComputer implementations not complete.

Unfortunately, it seems likely we will need all the binary map combinations, due to limitations in how converters work (they operate independently on each parameter, without knowledge of the other parameters).

> Avoid method calls within loops.

In the example, typically there is no difference on modern JVMs, but it should be tested. If it's easy to factor out, there is no harm, though.

> not `II` for `IterableInterval`

OK, let's use `II`. But if you rename it, please do it everywhere. We will also have to think about `RealRandomAccessibleRealInterval`—it hasn't come up yet but should probably be `RRARI`, as ugly as that is.

**Comment by LeonYang5114** [comment]
@ctrueden For the naming of the map ops, I think I will also make a separate branch for creating sub package for the map, which is swelling now.

**Comment by LeonYang5114** [comment]
Addressing the last task, we might want a complete collection of Map ops. They should be:
    UnaryMaps: II_II, II_RAI, RAI_II
    BinaryMaps: II_II_II, II_II_RAI, II_RAI_II, RAI_II_II, II_RAI_RAI, RAI_II_RAI, RAI_RAI_II
(10 map ops since RAI_RAI and RAI_RAI_RAI are not necessary).
    InplaceMaps: Unary - II_II    Binary - II_II_II, II_RAI_II, RAI_II_RAI
Each map should have a parallel version, so we need (10 + 4) * 2 = 28 map ops in total to make the namespace complete.

It will definitely make sense to auto generate all these classes. To make this easier, a `MapUtils` class might also be necessary. It helps `conform` the map element producers (II and RAI), and do the iteration logics. I am also considering having a method to wrap an RAI given an `IterationOrder`. More discussion on this later.

**Comment by dietzc** [comment]
>  I am also considering having a method to wrap an RAI given an IterationOrder 

how would you do that? At the moment this is only possible for `FlatIterationOrder` using `Views.flatIterable(...)`. Anyway, it's really important to get the priorities of the `Map`s right and understand what we need and what we're doing. For example, there may exists `Converter`s from `RandomAccessibleInterval` to `IterableInterval`. So we actually never need the `RAI` to `RAI` map explicitly.  So first check will always be: `are you `IterableInterval`s and if so, do you have the same `IterationOrder` and then it starts getting tricky: What if the `IterableInterval` is of type `FlatIterationOrder` and the other is also a `RAI` (but still a `IterableInterval` with non-flat `IterationOrder`? Then we can convert the `RAI` to a `IterableInterval` with `FlatIterationOrder` etc. etc. etc.

Again, this is really not too easy. Also, I'm not sure if the matching is already powerful enough to cover some of our use-cases.

Anyway, thanks for your great work!! :+1:

Btw: There are situation where we really can't perform a `Map`, e.g. on two `IterableInterval`s which are not `RAI`s and differ in `IterationOrder`.

**Comment by ctrueden** [comment]
I spoke with @LeonYang5114 and our current plan is to create a couple of helper cursors, maybe something like `LockedCursor1`/`LockedCursor2` or `SyncedCursor1`/`SyncedCursor2`.

The `LockedCursor1` would take an `IterableInterval` and a `RandomAccessibleInterval`, and would internally call `localizingCursor()` on the `IterableInterval`. When `fwd()` is called, it would call `fwd()` on its wrapped localizing `Cursor` as well as call `setPosition` on the RAI to match. Same for `jumpFwd()` etc., keeping the RAI position in sync with the cursor. And then it would have getters for both the II value and the RAI value at the current position.

The `LockedCursor2` (or whatever it's named) would do the same, but for two RAIs. The reason to have the two classes separate is to preserve the generics, so we have full type safety for the getters.

The `Colocalization_Analysis` project actually has a similar thing called [TwinCursor](https://github.com/fiji/Colocalisation_Analysis/blob/Colocalisation_Analysis-2.2.0/src/main/java/net/imglib2/TwinCursor.java) that does this sort of thing between two `Cursor` instances. We can migrate that into Ops or ImgLib2 core, as well, as needed.

Does this sound OK to you @dietzc?

**Comment by LeonYang5114** [comment]
@ctrueden @dietzc 

I was considering another way to achieve the goal of iterating RAIs and IIs, which require some structural refactoring, but I think that looks pretty cool.

Problem:
consider three methods `map(II, II, RAI, biOp), map(II, RAI, II, biOp), map(RAI, II, II, biOp)`, all they do is iterate the IIs and RAIs and call the `compute2` with the elements retrieved from the input1, input2, and output. The code for the first method looks like this:
```
	public static <I1, I2, O> void map(final IterableInterval<I1> a,
		final IterableInterval<I2> b, final RandomAccessibleInterval<O> c,
		final BinaryComputerOp<I1, I2, O> op)
	{
		final Cursor<I1> aCursor = a.localizingCursor();
		final Cursor<I2> bCursor = b.cursor();
		final RandomAccess<O> cAccess = c.randomAccess();
		while (aCursor.hasNext()) {
			aCursor.fwd();
			cAccess.setPosition(aCursor);
			op.compute2(aCursor.get(), bCursor.next(), cAccess.get());
		}
	}
```

Analysis:
It is easy to realize that the three `map`s have very similar structure: in the second `map(II a, RAI b, II c, BiOp op)`, we can actually permute the parameters to call the first `map(a, c, b, op)`, as long as we can somehow "tell" the first `map` to also permute the element parameters when calling `compute2()`.

Proposal:
- Create `BinaryComputer` interface, which is anotated with `@FunctionalInterface` and has one method `compute2(I1 inpu1, I2 input2, O output)`
- Make `BinaryComputerOp` extends `BinaryComputer`, and change all usage of `BinaryComputerOp` to `BinaryComputer` so long as the usage only cares about the `compute2()` method.
- Use lambda expression to pass a `BinaryComputer` that has its parameters permuted.

Summary:
Call ```MapUtils.map(a, c, b, (x, y, z) -> op.compute2(x, z, y));``` in the `map(II a, RAI b, II c, BiOp op)` method. This should use the logic of the first `map`.

I find it hard to express this idea clearly. I will push a new commit to show it. Sorry about the confusion.

**Comment by ctrueden** [comment]
> I think it also help separate the `computeX()` method from other methods

Definitely.

**Comment by dietzc** [comment]
> The LockedCursor1 would take an IterableInterval and a RandomAccessibleInterval, and would internally call localizingCursor() on the IterableInterval. When fwd() is called, it would call fwd() on its wrapped localizing Cursor as well as call setPosition on the RAI to match. Same for jumpFwd() etc., keeping the RAI position in sync with the cursor. And then it would have getters for both the II value and the RAI value at the current position.

Sounds nice, could also be something which goes into ImgLib2-core. However, I don't see how this helps us at the moment? We basically do exactly this in the current `Map` implementation. The problem I described is more on the `Converter` and `OpMatcher` side.

**Comment by LeonYang5114** [comment]
#337 is ready now.

**Comment by ctrueden** [comment]
I merged #337, so I am closing this issue. I think all points above are addressed except "Avoid method calls within loops." The JIT is actually very good at optimizing method calls these days, so I would not necessarily worry about method calls within loops. Better to [avoid premature optimization](http://programmers.stackexchange.com/q/80084/213866) since doing so avoids unnecessary complexity.
